### DISCUSSION TITLE
[Question] in lecture 8 pdf

### DISCUSSION THREAD

**Discussion by sms0509** [prosyslab-classroom/cs424-program-reasoning#330] [Q&A]
On the second page of lecture 8 slides, it is written that 

'Fully automated verification: combined with automated invariant generation methods', 

and I think I am not fully understanding the meaning of this sentence. Where the automated invariant is generated when verifying the program via constraint solving? The lecture 8 treated about CHC, and this method of verification uses just checking the satisfiability of FOL formula, not about inductive invariant and Hoare logic. What part do I misunderstand?

Thank you.

**Comment by KihongHeo** [comment]
Hi. Good question.

Checking the reachability to the error state requires inductive invariants.

See the example on page 11 in `lecture8.pdf`.  In this case, how do you check the reachability to the error state (more concretely, `Else(x)` for some x)?

To check the reachability, we need an invariant. Here, the following set is the invariant we want to find:
```
Entry
Cond(10), Cond(9), Cond(8), Cond(7), ...,  (i.e., for all i <= 10. Cond(i))
Body(9), Body(8), Body(7), ...,  (i.e., for all i < 10. Body(i))
End(10)
Else(10)
 ```
Why this is invariant? If you apply the set to the rules again, you will get the same set.

Then how is it related to the Hoare Logic? The above set indicates the following Hoare triples are valid:
$\lbrace \top \rbrace$ `code in Entry` $\lbrace x <= 10 \rbrace$
$\lbrace x <= 10 \rbrace$ `code in Cond and Body` $\lbrace x = 10 \rbrace$
(Of course, there exists some mismatch between C and LLVM)

Notice that `code in Cond and Body` corresponds to the while loop in the C program. Such invariants can be computed via CHC solving.

**Comment by sms0509** [comment]
Thank you! I got it.
